Added CRS indexing based on an sqlite database

gktval opened the pull request:
Implemented a database service to obtain fast indexing of coordinates systems by codes from an sqlite database. Also added a file service to obtain indexing coordinates systems from a csv file. 

Fixed parsing errors of wkts.
Added and EXTENSION parse.


There are now three CoordinateSystemServices:
1) Default
2) File (this is based on a csv. I initially used this method to test everything) You can obtain this file by exporting the table in the db. Nevertheless, if a user wants to provide a comma delimited file with the same headers as the table, they can.
3) Database 

I am not too sure about how to instantiate the database... 
Previously, when CoordinateSystemServices is called without any parameters, it would initialize the Default CRS dictionary of WGS84 and TransMercator. I didn't think this was extremely viable (since it was only two coordinate systems), so I made the default constructor to call the database service. If you want to call the Default service, you should provide any enumeration (or null). 

The database was stripped down a bit. For example, about 30 codes overlap with ESRI and EPSG. I removed the ESRI duplicates from the database so that I could make the PrimaryKey the code.  I also remove the IGNF authority because these had string values as the code (instead of integer).

The only known problem that persists is parsing horizontal CS with more than two axis. I had commented this exception out while testing. Otherwise, all of the wkts in the database should parse.

Commit:
Implemented a more dynamic CoordinateSystemService

Implemented a database service to obtain fast indexing of coordinates systems by codes from an sqlite database. Also added a file service to obtain indexing coordinates systems from a csv file.
Fixed parsing errors of wkts and added and EXTENSION parse.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/NetTopologySuite/ProjNet4GeoAPI?pullRequest=108) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/NetTopologySuite/ProjNet4GeoAPI?pullRequest=108) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: gktval<br/>:x: Travis Yeik<br/><hr/>**Travis Yeik** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/NetTopologySuite/ProjNet4GeoAPI?pullRequest=108) it.</sub>

FObermaier commented on src/ProjNet/Services/ICoordinateSystemService.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace ProjNet.Services
+{
+    public interface ICoordinateSystemService
```
A public interface requires proper xml-code documentation

FObermaier commented on src/ProjNet/Services/DatabaseCoordinateService.cs:
```diff
@@ -0,0 +1,81 @@
+    /// <summary>
+    /// Coordinate System Service tied to a database
+    /// </summary>
+    internal class DatabaseCoordinateService : ICoordinateSystemService
```
Move in separate project (database too), make it public there to be instantiable.
This will keep core ProjNet nuget package small

FObermaier commented on src/ProjNet/ProjNET.csproj:
```diff
@@ -18,14 +18,27 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="sqlite-net-pcl" Version="1.8.116" />
```
move to sperate project, see DatabaseCoordinateService.cs

FObermaier commented on src/ProjNet/ProjNET.csproj:
```diff
@@ -18,14 +18,27 @@
+    <Compile Remove="IO\SqlReaderHelper.cs" />
+  </ItemGroup>
+
+  <ItemGroup>
```
move to sperate project, see DatabaseCoordinateService.cs

FObermaier commented on src/ProjNet/IO/DatabaseProvider.cs:
```diff
@@ -0,0 +1,100 @@
+
+namespace ProjNet.IO
+{
+    public class DatabaseProvider
```
move to sperate project, see DatabaseCoordinateService.cs

FObermaier commented on src/ProjNet/CoordinateSystemServices.cs:
```diff
@@ -14,94 +14,26 @@
-
-        #endregion
-
-        /// <summary>
```
This is a breaking change, uncomment and mark [Obsolete]

FObermaier commented on src/ProjNet/IO/CoordinateSystems/WKTStreamTokenizer.cs:
```diff
@@ -169,5 +169,18 @@ public void ReadAuthority(out string authority, out long authorityCode)
             ReadCloser(bracket);
         }
+
+        public void ReadExtension(out string extension, out string extensionMethod)
```
Separate PR would be great

FObermaier reviewed (changes requested):
Thanks for your input @gktval, really appreciated.

It'd b great if you could create a separate PR for the ReadExtension so we can map it to #106

gktval commented:
I will separate out the EXTENSION. I am still not sure about the the sqlite database being part of the project. So I think I will redo it as well. I think that it is a great addition and rounds the project off, but I don't know how I feel about forcing people to download 6Mb of data if they don't want to use it. I think there are two options: 1) we leave it as a separate download on this github repository and people download it separately, then pass the connection string as a parameter or 2) the sqlite database, the DatabaseService, and the sqlite-net-pcl DLL is put a separate project that people can attach to ProjNet through the ICoordinateSystemService. Thoughts?

Oh nevermind, I see your comments above. We are on the same page.

Commit:
Separated out sqlite methods. Added NUTest

gktval commented:
I separated out the sql stuff to ProjNet.Sqlite. I also added unit test for each of the services. It looks a lot cleaner now. 

There are two EXTENSIONS, one to the Crs and one to the projection. Both of those are added. I really have no idea how to create a separate PR for only the extension stuff; is it possible to close #106 with this PR?

FObermaier commented:
@gktval, I managed to seperate out EXTENSION handling bits or this PR. I see some issues with the code that need to be resolved. Please see https://github.com/NetTopologySuite/ProjNet4GeoAPI/pull/111

Commit:
Changes to ` Add proper handling of EXTENSION tag #111 `
Removed the AddParameter method.
Added async search methods to the database.

Commit:
almost ready

Commit:
Quickmarks

Commit:
update

gktval committed:
Merge branch 'develop' into develop

Commit:
update csproj file

Update for compatibility with Maui in iOs and Android (ie, embedded resources). Create the Database file if it cannot be found so that things don't crash.

Commit:
Merge branch 'develop' of https://github.com/gktval/ProjNet4GeoAPI into develop
